### ISSUE TITLE
Add support for "in" builtin test

### ISSUE THREAD

**Issue by keirlawson** [Keats/tera#405]
It would be great if Tera could support testing whether a given variable is equals to one of a series of values, ie Jinja2's `in` test: http://jinja.pocoo.org/docs/2.10/templates/#in

**Comment by Keats** [comment]
That would be very easy to add if you're interested in contributing or I'll add it later (on the v1 branch)

**Comment by keirlawson** [comment]
I'd be interested in having a go at contributing this, any pointers for where I should start?

**Comment by Keats** [comment]
First thing is to do the PR on the `v1` branch and then add a function + test in https://github.com/Keats/tera/blob/master/src/builtins/testers.rs
Once this is done, you only need to add it in https://github.com/Keats/tera/blob/v1/src/tera.rs#L568 and that should be done

**Comment by keirlawson** [comment]
I've had an initial stab at it, see https://github.com/Keats/tera/compare/v1...keirlawson:v1 - however I am getting the following error when trying to run the example:

```
$ cargo run --example basic
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
     Running `target/debug/examples/basic`
Parsing error(s):
* Failed to parse "examples/basic/templates/base.html"
  --> 10:16
   |
10 | {% if username in ["Bob", "Cathy"] %}␊
   |                ^---
   |
   = expected `or`, `and`, `<=`, `>=`, `<`, `>`, `==`, `!=`, `+`, `-`, `*`, `/`, `%`, a filter, or `%}` or `-%}`
```

Any thoughts on what I'm missing?

**Comment by Keats** [comment]
Ah that looks like it's stuck at the parsing level, I'll have a look after. It had many changes recently so I wouldn't be suprised if there was an issue with ti

**Comment by Keats** [comment]
Ah the grammar expects the `is` keyword so the actual call will be `{% if username is in(["Bob", "Cathy"]) %}`
Not as pretty I agree

**Comment by keirlawson** [comment]
With your suggested syntax I get

```
Parsing error(s): 
* Failed to parse "examples/basic/templates/base.html"
  --> 10:22
   |
10 | {% if username is in(["Bob", "Cathy"]) %}␊
   |                      ^---
   |
   = expected a value that can be negated
```

Would it be possible to change the parser to allow this to more closely match jinja2? I have a load of jinja2 templates with which Tera is nearly compatible, bar this one builtin...

**Comment by Keats** [comment]
Ah that one is an easy fix, I have half done already but no time to finish it today.
Afaik the only difference are the parentheses right?

**Comment by keirlawson** [comment]
no, you also don't need `is` with jinja2

**Comment by Keats** [comment]
You're thinking of an if expression. Jinja does a `in` test that works exactly like the last sample you've posted.
The code that is doing `{% if username in ["Bob", "Cathy"] %}` is not a test, it's just Python code that gets executed: it is the standard Python `in` keyword there.

**Comment by keirlawson** [comment]
Ah your right, looks like that is distinct from tests.  I don't think its just arbritary python though, it appears Jinja2 has it as an explicitly codified operators as well as a test: http://jinja.pocoo.org/docs/2.10/templates/#other-operators - so perhaps we could implement an `in` operator in Tera in addition to the test builtin?

**Comment by Keats** [comment]
Ah I didn't realise that. Looks like the test was added for some funky things the operator couldn't do: https://github.com/pallets/jinja/pull/658

First thing to do is to add array support for tests arguments which is easy and I've already done half yesterday.
Then to decide whether to add support for the `in` operator, which is another big issues on its own. Aiming for compatibility is good but it will never be fully compatible, for example there is no dict literal in Tera right now, nor is it possible to call a function (obviously).

**Comment by keirlawson** [comment]
Ok, I guess I can finish up the PR for the `in` builtin, and open up a new issue for the `in` operator?

**Comment by Keats** [comment]
Looking at the Jinja2 issue it doesn't look like the test adds anything over the operator in Tera itself since we don't have the filters mentioned in it so I would probably not include it if we add the operator, at least for now.

**Comment by keirlawson** [comment]
Ok, let me know if there's anything it makes sense for me to pick up for this. FWIW my use case is only with arrays, not dicts

**Comment by Keats** [comment]
> FWIW my use case is only with arrays, not dicts

But people are definitely going to ask for it :p
I'll think about it over the weekend. The `in` operator without dict literals should be easy enough to implement.

**Comment by Keats** [comment]
@keirlawson can you play with https://github.com/Keats/tera/pull/408 a bit?

**Comment by keirlawson** [comment]
Slotted it into my project and works exactly as I would expect, thanks for this :)

**Comment by Keats** [comment]
Nice! I'll release a new beta in a couple of days.
